### PR TITLE
Use subsystem listener object to provide params for SPDK calls

### PR DESCRIPTION
**Pull request by artek-koltun**

### CODE REVIEW

File: pkg/frontend/frontend.go [hunk 1]
```diff
@@ -18,32 +21,83 @@ var (
 	errUnexpectedSpdkCallResult = status.Error(codes.FailedPrecondition, "Unexpected SPDK call result.")
 )
 
+// SubsystemListener interface used to provide SPDK call params to create/delete
```
Comment (intelfisz) [inline pkg/frontend/frontend.go:24]: is used

File: pkg/frontend/frontend.go [hunk 2]
```diff
@@ -18,32 +21,83 @@ var (
...
 
-	rpc  server.JSONRPC
-	Nvme NvmeParameters
+	rpc            server.JSONRPC
+	subsysListener SubsystemListener
+	nvme           *NvmeParameters
 }
 
+// Option defines and optional argument at frontend server creation.
+type Option func(*Server) *Server
```
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/frontend/frontend.go:58]: Added note which functions can be used to modify arguments in `NewServerWithJSONRPC` doc

File: pkg/frontend/frontend.go (hunk 1 above)
Comment (artek-koltun) [reply pkg/frontend/frontend.go]: done

File: pkg/frontend/frontend.go [hunk 3]
```diff
@@ -18,32 +21,83 @@ var (
...
+
 // Server contains frontend related OPI services
 type Server struct {
 	pb.UnimplementedFrontendNvmeServiceServer
 	pb.UnimplementedFrontendVirtioBlkServiceServer
 	pb.UnimplementedFrontendVirtioScsiServiceServer
 
-	rpc  server.JSONRPC
-	Nvme NvmeParameters
+	rpc            server.JSONRPC
```
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/frontend/frontend.go:52]: Added note which functions can be used to modify arguments in NewServerWithJSONRPC doc

File: pkg/frontend/frontend.go (hunk 1 above)
Comment (artek-koltun) [reply pkg/frontend/frontend.go]: done

File: pkg/frontend/frontend.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (artek-koltun) [reply pkg/frontend/frontend.go:58]: added a note

File: pkg/frontend/frontend.go [hunk 4]
```diff
@@ -59,4 +74,27 @@ func NewServer(jsonRPC server.JSONRPC) *Server {
 			ScsiLuns:  make(map[string]*pb.VirtioScsiLun),
 		},
 	}
+
+	for _, option := range options {
+		server = option(server)
```
Comment (glimchb) [inline pkg/frontend/frontend.go:79]: why create a new server `server = option(server)`  here ? is it to chain calls ? like `optionA(optionB(server))` ? If so, need to document this

**Review by glimchb — CHANGES_REQUESTED** [review]
let's split into smaller commits... then we can review it
- I understand you want to create and `Option` interface to provide functions during server creation
  - if that's true that should be separate commit across all servers and not just frontend
  - inside c-tor you call them one by one and responsibility of the function to add/change something to the server and those functions are not saved in the server by themselves...
  - Maybe it is limiting to have something called once in c-tor and not every time? 
 - Then second commit could be specific use of this mechanism for SubsystemListener interface 
   - you create the iface, then the struct and c-tor and finally the Params function with actual impl
   - So looks like for every option you need to create a corresponding object inside the server...

The question is: can we just pass the function `GetSubsystemAddListenerParams()` to server in c-tor, save it in server and call it during `CreateNVMeSubsystem` grpc ?

I guess I'm trying to simplify things here...

**Comment by artek-koltun** [comment]
> let's split into smaller commits... then we can review it
> 
>     * I understand you want to create and `Option` interface to provide functions during server creation
>       
>       * if that's true that should be separate commit across all servers and not just frontend
>       * inside c-tor you call them one by one and responsibility of the function to add/change something to the server and those functions are not saved in the server by themselves...
>       * Maybe it is limiting to have something called once in c-tor and not every time?
> 
>     * Then second commit could be specific use of this mechanism for SubsystemListener interface
>       
>       * you create the iface, then the struct and c-tor and finally the Params function with actual impl
>       * So looks like for every option you need to create a corresponding object inside the server...
> 
> 
> The question is: can we just pass the function `GetSubsystemAddListenerParams()` to server in c-tor, save it in server and call it during `CreateNVMeSubsystem` grpc ?
> 
> I guess I'm trying to simplify things here...

1. Not sure that we need to propagate `Option` across all other packages if they do not use it. As a probable simplification, I can provide a dedicated constructor method `NewServerWithSubsystemListener` what do you think?
2. I think the interface is preferable here. I expect we need to extend it later for vfiouser transport since it requires a dedicated folder creation/delition per controller

**Review by glimchb — APPROVED** [review]
Thanks for reworking. Looks much better!

File: pkg/frontend/frontend.go [hunk 5]
```diff
@@ -60,3 +71,14 @@ func NewServer(jsonRPC server.JSONRPC) *Server {
 		},
 	}
 }
+
+// NewServerWithSubsystemListener creates initialized instance of FrontEnd server communicating
+// with provided jsonRPC and externally created SubsystemListener instead default one.
+func NewServerWithSubsystemListener(jsonRPC server.JSONRPC, sysListener SubsystemListener) *Server {
+	if sysListener == nil {
+		log.Panic("nil for SubsystemListener is not allowed")
+	}
```
Comment (glimchb) [inline pkg/frontend/frontend.go:80]: What do you think about panic vs return err here ?

Comment (artek-koltun) [reply pkg/frontend/frontend.go:80]: This method is used only at wiring stage when we start our server and we should clearly indicate that something is going wrong and the system will be inconsistent. Error can be ignored, but panic requires reactions.

**Pull request merged**